src/aiortc/contrib/media.py

```diff
@@ -79,6 +80,89 @@ async def stop(self):
         self.__tracks = {}
 
 
+class BroadcasterStreamTrack(MediaStreamTrack):
+    def __init__(self, broadcaster, source: MediaStreamTrack) -> None:
+        super().__init__()
+        self.kind = source.kind
+        self._broadcaster = broadcaster
+        self._queue: asyncio.Queue[Optional[Frame]] = asyncio.Queue()
+        self._source: Optional[MediaStreamTrack] = source
+
+    async def recv(self):
+        if self.readyState != "live":
+            raise MediaStreamError
+
+        self._broadcaster._start(self)
+        frame = await self._queue.get()
+        if frame is None:
+            self.stop()
+            raise MediaStreamError
+        return frame
+
+    def stop(self):
+        super().stop()
+        if self._broadcaster is not None:
+            self._broadcaster._stop(self)
+            self._broadcaster = None
+            self._source = None
+
+
+class MediaBroadcaster:
+    """
+    A media source that relays one or more tracks.
+    """
+
+    def __init__(self) -> None:
+        self.__proxies: Dict[MediaStreamTrack, Set[BroadcasterStreamTrack]] = {}
+        self.__tasks: Dict[MediaStreamTrack, asyncio.Future[None]] = {}
+
+    def subscribe(self, track: MediaStreamTrack) -> MediaStreamTrack:
+        """
+        Create a proxy around the given `track` for a new consumer.
+        """
+        proxy = BroadcasterStreamTrack(self, track)
+        self.__log_debug("Create proxy %s for source %s", id(proxy), id(track))
+        if track in self.__proxies:
+            self.__proxies[track].add(proxy)
+        else:
+            self.__proxies[track] = set([proxy])
+        return proxy
+
+    async def _run(self, track: MediaStreamTrack) -> None:
+        while True:
+            try:
+                frame = await track.recv()
+            except MediaStreamError:
+                frame = None
+            for proxy in self.__proxies[track]:
+                proxy._queue.put_nowait(frame)
+            if frame is None:
+                return
+
+    def _start(self, proxy: BroadcasterStreamTrack) -> None:
+        track = proxy._source
+        if track is not None and track not in self.__tasks:
+            self.__log_debug("Start reading source %s" % id(track))
+            self.__tasks[track] = asyncio.ensure_future(self._run(track))
+
+    def _stop(self, proxy: BroadcasterStreamTrack) -> None:
+        track = proxy._source
+        if track is not None:
+            self.__log_debug("Discard proxy %s", id(proxy))
+            self.__proxies[track].discard(proxy)
+
+            # if there are no subscribers left, stop reading the source
+            if not self.__proxies[track]:
+                del self.__proxies[track]
+                if track in self.__tasks:
+                    self.__log_debug("Stop reading source %s", id(track))
+                    self.__tasks[track].cancel()
```

Comment:
Correct me if I am wrong, but in case of adding this broadcaster at the receiving side:

- Queue of RemoteStreamTrack would be able to be filling up before any proxy connected to broadcaster

- Disconnecting all streams from broadcaster, stops keep emptying the queue of RemoteStreamTrack. Reconnecting any proxy track would seem like the received stream was 'paused'.

Reply:
Yeah, we should probably keep reading and discarding frames if we don't want to accumulate stale frames. This makes me wonder what the correct trigger is to stop reading completely. Probably an explicit `stop()` corouting, because our MediaBroadcast is a hybrid between a source and a sink.

Reply:
.. or we just let the user call `stop()` on the source track when they are done with it, this will make the `_run()` coroutine stop, and we can perform our cleanup there.